src/mq/gcloud-pubsub-push/mq.ts

```diff
@@ -0,0 +1,433 @@
+                    { fedifyId, pubSubId: message.id },
+                );
+            } catch (error) {
+                message.nack();
+
+                this.logger.error(
+                    `Failed to handle message [FedifyID: ${fedifyId}, PubSubID: ${message.id}]: ${error}`,
+                    { fedifyId, pubSubId: message.id, error },
+                );
+
+                this.events.emit(this.errorEvent, error);
+            }
+        });
+
+        // Return a promise that resolves when the message queue is stopped
+        return await new Promise((resolve) => {
+            options.signal?.addEventListener('abort', () => {
+                this.events.removeAllListeners();
+                resolve();
+            });
+        });
+    }
+
+    /**
+     * Handles a message
+     *
+     * Ensure that the message queue is listening before calling this method
+     *
+     * @param message {Message} Message to handle
+     */
+    handleMessage(message: Message): void {
+        if (this.isListening === false) {
+            this.logger.warn(
+                `Message [FedifyID: ${message.attributes.fedifyId}, PubSubID: ${message.id}] cannot be handled as the message queue is not yet listening`,
+                { fedifyId: message.attributes.fedifyId, pubSubId: message.id },
+            );
+        }
+
+        this.events.emit(this.messageEvent, message);
+    }
+
+    /**
+     * Registers an error listener
+     *
+     * If an error occurs, the provided listener will be executed with the error
+     *
+     * @param listener {function}
+     */
+    registerErrorListener(listener: (error: Error) => void): void {
+        this.events.on(this.errorEvent, listener);
+    }
+}
+
+/**
+ * Represents an incoming message from a Pub/Sub push subscription
+ *
+ * @see https://cloud.google.com/pubsub/docs/push#receive_push
+ */
+interface IncomingPushMessageJson {
+    message: {
+        /**
+         * Unique identifier for the message
+         */
+        message_id: string;
+        /**
+         * Data contained within the message encoded as a base64 string
+         */
+        data: string;
+        /**
+         * Additional metadata about the message
+         */
+        attributes: Record<string, string>;
+    };
+}
+
+/**
+ * Hono middleware to handle an incoming message from a Pub/Sub push subscription
+ *
+ * @param mq {GCloudPubSubPushMessageQueue} Message queue instance
+ * @returns {function}
+ *
+ * @example
+ * ```
+ * import { createMessageQueue, handlePushMessage } from './mq/gcloud-pubsub-push';
+ *
+ * const queue = await createMessageQueue(...);
+ *
+ * app.post('/mq', handlePushMessage(queue));
+ * ```
+ */
+export function handlePushMessage(
+    mq: GCloudPubSubPushMessageQueue,
+): (ctx: Context) => Promise<Response> {
+    return async (ctx: Context) => {
+        const json = await ctx.req.json<IncomingPushMessageJson>();
```

Comment:
I don't think we can declare the `IncomingPushMessageJson` type here - as we've no idea what shape the data is? I think this should be `unknown` and either have manual type checks, or use `zod` to validate the type like we do for webhooks